### PR TITLE
Implementing RFC 1002: workspace archiving and deletion

### PR DESCRIPTION
**Pull request by vxgmichel**
Implementing [RFC 1002 for workspace archiving and deletion](https://github.com/Scille/parsec-cloud/pull/5050)

### CODE REVIEW

File: parsec/backend/memory/block.py [hunk 1]
```diff
@@ -69,6 +71,11 @@ def _check_realm_access(
             realm = self._realm_component._get_realm(organization_id, realm_id)
```
Comment (touilleMan) [inline parsec/backend/memory/block.py:71]: should pass now here

File: parsec/backend/memory/realm.py [hunk 2]
```diff
@@ -80,6 +80,11 @@ def is_archived(self) -> bool:
             return False
         return self.last_archiving_configuration_request.configuration.is_archived()
 
+    def is_deletion_planned(self) -> bool:
```
Comment (touilleMan) [inline parsec/backend/memory/realm.py:83]: add a comment to precise deleted is *also* deletion planned !

File: parsec/backend/postgresql/organization.py [hunk 3]
```diff
@@ -485,3 +488,23 @@ async def update(
 
             if with_is_expired and is_expired:
                 await send_signal(conn, BackendEvent.ORGANIZATION_EXPIRED, organization_id=id)
+
+    async def archiving_config(
+        self, id: OrganizationID, user: UserID
+    ) -> list[RealmArchivingStatus]:
+        async with self.dbh.pool.acquire() as conn:
+            await self._get(conn, id)
+        realms = await realm_queries.query_get_realms_for_user(conn, id, user)
```
Comment (touilleMan) [inline parsec/backend/postgresql/organization.py]: use conn out of the context manager ?

**Review by touilleMan — CHANGES_REQUESTED** [review]

File: docs/rfcs/1002-workspace-archiving-and-deletion.md [hunk 4]
```diff
@@ -32,33 +32,13 @@ More specifically, the following API changes are required:
...
-                ]
-            }
-        ]
+```rust
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Hash, Serialize, Deserialize)]
+#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
+pub enum RealmArchivingConfiguration {
+    Available,
+    Archived,
+    DeletionPlanned(DateTime),
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply docs/rfcs/1002-workspace-archiving-and-deletion.md:41]: 986b01dedbdbc4ea21f341eb7621050ba0015b33

[the remark this answers is not in the document]
Comment (vxgmichel) [reply docs/rfcs/1002-workspace-archiving-and-deletion.md:41]: 76cabe38c738d75747f09d416edb0260bd079d69

File: oxidation/libparsec/crates/protocol/schema/authenticated_cmds/organization_config.json5 [hunk 5]
```diff
@@ -29,6 +29,11 @@
                         "name": "sequester_services_certificates",
                         "type": "RequiredOption<List<Bytes>>",
                         "introduced_in": "2.8"
+                    },
+                    {
+                        "name": "minimum_archiving_period",
+                        "type": "Integer", // In seconds
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply oxidation/libparsec/crates/protocol/schema/authenticated_cmds/organization_config.json5]: 4fb391478c504b03c953ac7566ae4f86b0f9128d

File: parsec/backend/memory/realm.py [hunk 6]
```diff
@@ -43,6 +47,9 @@ class Realm:
     checkpoint: int = attr.ib(default=0)
     granted_roles: List[RealmGrantedRole] = attr.ib(factory=list)
     last_role_change_per_user: Dict[UserID, DateTime] = attr.ib(factory=dict)
+    last_archiving_configuration_request: RealmArchivingConfigurationRequest | None = attr.ib(
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply parsec/backend/memory/realm.py]: 1c0f8dfc6d4f77ccce3383a555699fad03588d78

[the remark this answers is not in the document]
Comment (vxgmichel) [reply parsec/backend/memory/realm.py]: 1c0f8dfc6d4f77ccce3383a555699fad03588d78

File: parsec/api/protocol/types.py [hunk 7]
```diff
@@ -86,3 +88,41 @@ def _deserialize(self, value: object, attr: str, data: dict[str, object]) -> Api
...
+
+class RealmArchivingConfigurationField(Field[RealmArchivingConfiguration]):
+    def _serialize(
+        self, value: RealmArchivingConfiguration | None, attr: str, obj: object
+    ) -> str | dict[str, DateTime] | None:
+        if value is None:
+            return None
+        if value.is_archived() or value.is_available():
+            return value.str
+        return {value.str: value.deletion_date}
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply parsec/api/protocol/types.py]: 76cabe38c738d75747f09d416edb0260bd079d69

File: parsec/core/mountpoint/manager.py [hunk 8]
```diff
@@ -252,6 +256,13 @@ async def mount_workspace(
         if (workspace_id, timestamp) in self._mountpoint_tasks:
             raise MountpointAlreadyMounted(f"Workspace `{workspace_id.hex}` already mounted.")
 
+        if (workspace_id, timestamp) in self._mounting_events:
+            raise MountpointAlreadyMounted(f"Workspace `{workspace_id.hex}` is beening mounted.")
```
Comment (touilleMan) [inline parsec/core/mountpoint/manager.py:260]: is being

File: parsec/backend/memory/realm.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (vxgmichel) [reply parsec/backend/memory/realm.py]: ced88b9bb544b26479493567d5922fd2d93f5d6d

[the remark this answers is not in the document]
Comment (vxgmichel) [reply parsec/backend/memory/realm.py]: ced88b9bb544b26479493567d5922fd2d93f5d6d

File: parsec/backend/memory/block.py (hunk 1 above)
Comment (vxgmichel) [reply parsec/backend/memory/block.py]: 4bdf0e2407b0936afe3dfa7ab9b30161cddea744

File: parsec/backend/memory/realm.py [hunk 9]
```diff
@@ -267,6 +292,60 @@ async def update_roles(
...
+        # Check minimum archiving period
+        assert self._organization_component is not None
+        organization = await self._organization_component.get(organization_id)
+        if not archiving_configuration_request.is_valid_archiving_configuration(
+            organization.minimum_archiving_period
+        ):
+            raise RealmArchivingPeriodTooShortError()
+
+        # Timestamp should be greater than last role
+        if last_role.granted_on >= archiving_configuration_request.configured_on:
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply parsec/backend/memory/realm.py:370]: 80861ef47d5a778002fe5ed28e7de4f206d4201d
e97cb1c22

File: parsec/backend/postgresql/realm_queries/update_archiving.py [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+)
+
+_q_get_archiving_configuration_for_update = Q(
+    f"""
+SELECT configuration, deletion_date, certified_on
+FROM realm_archiving
+WHERE
+    realm = { q_realm_internal_id(organization_id="$organization_id", realm_id="$realm_id") }
+ORDER BY certified_on DESC LIMIT 1
+FOR UPDATE
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply parsec/backend/postgresql/realm_queries/update_archiving.py]: 80861ef47d5a778002fe5ed28e7de4f206d4201d

File: src/data/certif.rs [hunk 11]
```diff
@@ -740,6 +740,17 @@ impl RealmArchivingConfiguration {
             libparsec::types::RealmArchivingConfiguration::DeletionPlanned(_)
         )
     }
+
+    #[args(now = "None")]
+    fn is_deleted(&self, now: Option<DateTime>) -> bool {
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply src/data/certif.rs]: done

File: parsec/core/remanence_monitor.py [hunk 12]
```diff
@@ -90,6 +90,17 @@ def _on_sharing_updated(
...
             _start_remanence_manager(new_entry.id)
 
+    def _on_archiving_updated(
+        event: CoreEvent,
+        workspace_id: EntryID,
+        configuration: RealmArchivingConfiguration,
+        configured_on: DateTime | None,
+    ) -> None:
+        # No reader rights
+        if configuration.is_deleted(user_fs.device.time_provider.now()):
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply parsec/core/remanence_monitor.py]: done

File: parsec/core/mountpoint/manager.py [hunk 13]
```diff
@@ -433,27 +433,42 @@ async def mountpoint_manager_factory(
...
+        if previous_entry is None:
             if mount_on_workspace_shared:
                 mount_nursery.start_soon(mountpoint_manager.safe_mount, new_entry.id)
             return
 
+    def on_archiving_updated_event(
+        event: CoreEvent,
+        workspace_id: EntryID,
+        configuration: RealmArchivingConfiguration,
+        configured_on: DateTime | None,
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply parsec/core/mountpoint/manager.py:472]: done

File: parsec/core/mountpoint/manager.py [hunk 14]
```diff
@@ -433,27 +433,42 @@ async def mountpoint_manager_factory(
...
+    ) -> None:
+        # Archiving updated
+        if event == CoreEvent.ARCHIVING_UPDATED and configuration.is_deleted(
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply parsec/core/mountpoint/manager.py]: done

File: parsec/backend/postgresql/realm_queries/update_archiving.py [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+from parsec.backend.realm import (
+    RealmAccessError,
+    RealmArchivingConfigurationRequest,
+    RealmArchivingPeriodTooShortError,
+    RealmDeletedError,
+    RealmNotFoundError,
+    RealmRoleRequireGreaterTimestampError,
+)
+
+_q_get_user_role = Q(
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply parsec/backend/postgresql/realm_queries/update_archiving.py]: https://github.com/Scille/parsec-cloud/issues/5203
https://github.com/Scille/parsec-cloud/pull/5061/commits/f972b816214f3820cba2a9ce3bab42af2e98dbb5

File: parsec/backend/memory/realm.py (hunk 2 above)
Comment (vxgmichel) [reply parsec/backend/memory/realm.py:103]: done

File: oxidation/libparsec/crates/core/src/events.rs [hunk 16]
```diff
@@ -85,6 +88,7 @@ impl CoreEvent {
             CoreEvent::MountpointTrioDeadlockError => "mountpoint.trio_deadlock_error",
             CoreEvent::PkiEnrollmentsUpdated => "pki_enrollment.updated",
             CoreEvent::SharingUpdated => "sharing.updated",
+            CoreEvent::ArchivingUpdated => "sharing.updated",
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply oxidation/libparsec/crates/core/src/events.rs]: done

File: parsec/core/backend_connection/authenticated.py [hunk 17]
```diff
@@ -339,6 +349,10 @@ def status_exc(self) -> Exception | None:
     def cmds(self) -> BackendAuthenticatedCmds | RsBackendAuthenticatedCmds:
         return self._cmds
 
+    @property
+    def all_monitors_task_statuses(self) -> list[MonitorTaskStatus]:
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply parsec/core/backend_connection/authenticated.py]: done

File: parsec/core/fs/userfs/userfs.py [hunk 18]
```diff
@@ -950,6 +978,68 @@ async def workspace_share(
         elif not isinstance(rep, RealmUpdateRolesRepOk):
             raise FSError(f"Error while trying to set vlob group roles in backend: {rep}")
 
+    async def update_archiving_status(self) -> DateTime | None:
+        rep = await self.backend_cmds.archiving_config()
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply parsec/core/fs/userfs/userfs.py]: done

File: parsec/core/fs/userfs/userfs.py [hunk 19]
```diff
@@ -950,6 +978,68 @@ async def workspace_share(
...
+
+        # Loop over workspaces
+        for status in rep.archiving_config:
+            workspace_id = EntryID.from_bytes(status.realm_id.bytes)
+            previous_configuration, previous_configured_on = self._archiving_configuration[
+                workspace_id
+            ]
+
+            # Compare previous and current state
+            previous_state = (
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply parsec/core/fs/userfs/userfs.py]: done

File: parsec/core/fs/userfs/userfs.py [hunk 20]
```diff
@@ -950,6 +978,68 @@ async def workspace_share(
...
+            )
+            if previous_state == current_state:
+                continue
+
+            # Update state and send events
+            self._archiving_configuration[workspace_id] = (
+                status.configuration,
+                status.configured_on,
+            )
+            self.event_bus.send(
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply parsec/core/fs/userfs/userfs.py]: done

File: parsec/core/sync_monitor.py [hunk 21]
```diff
@@ -415,6 +416,23 @@ def _on_sharing_updated(
                 ctx.due_time = user_fs.device.timestamp().timestamp()
                 _trigger_early_wakeup()
 
+    def _on_archiving_updated(
+        sender: str,
+        workspace_id: EntryID,
+        configuration: RealmArchivingConfiguration,
+        configured_on: DateTime | None,
+    ) -> None:
+        # If role have changed we have to reset the sync context given
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply parsec/core/sync_monitor.py]: done

**Review by touilleMan — APPROVED** [review]

**Pull request merged**